### PR TITLE
Make README more useful

### PR DESCRIPTION
**Pull request by rajkan01**
* Provide description on application functionality
* Add official example badge
* Remove outdated information
* Link to Mbed boards
* Indicate how to find out which toolchain support a target chosen
* Provide instructions to flash after building
* Updated the tests README to give proper information with example

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,59 +1,94 @@
...
-| mbed-os\platform |  3657 |   104 |  604 |
-| mbed-os\rtos     | 13822 |  2310 | 4592 |
-| mbed-os\targets  |  9193 |   104 |  324 |
-| Subtotals        | 40397 |  2564 | 7260 |
-+------------------+-------+-------+------+
-Total Static RAM memory (data + bss): 9824 bytes
-Total Flash memory (text + data): 42961 bytes
+ID:            Thread id
+Name:          Thread name
+State:         Thread state
```
Comment (evedon) [inline README.md]: Can you give indication on what State 1, 2, 3 means (see osThreadState_t)

File: README.md [hunk 2]
```diff
@@ -1,59 +1,94 @@
...
+Stack Space: 448 
 
-Your PC may take a few minutes to compile your code. At the end, you see the following result:
+ID: 0x20000f14 
+Name: rtx_timer 
+State: 3 
+Priority: 40 
+Stack Size: 768 
+Stack Space: 664
+```
```
Comment (evedon) [inline README.md:83]: In the example, could you start a few threads so we see more statistics (perhaps threads in the other states)?

File: README.md [hunk 3]
```diff
@@ -1,59 +1,94 @@
...
+* [Mbed OS configuration](https://os.mbed.com/docs/latest/reference/configuration.html).
+* [Mbed OS serial communication](https://os.mbed.com/docs/latest/tutorials/serial-communication.html).
+* [Mbed boards](https://os.mbed.com/platforms/).
 
 ### License and contributions
-The software is provided under Apache-2.0 license. Contributions to this project are accepted under the same license. Please see contributing.md for more info.
 
-This project contains code from other projects. The original license text is included in those source files. They must comply with our license guide
\ No newline at end of file
+The software is provided under the Apache-2.0 license. Contributions to this project are accepted under the same license. Please see contributing.md for more info.
```
Comment (evedon) [inline README.md]: for more information (instead of for more info)

File: README.md [hunk 4]
```diff
@@ -1,59 +1,94 @@
...
+This guide reviews the steps required to get thread statistics on an Mbed OS enabled platform.
 
-Please install [mbed CLI](https://github.com/ARMmbed/mbed-cli#installing-mbed-cli).
+You can build this project with all supported [Mbed OS build tools](https://os.mbed.com/docs/mbed-os/latest/tools/index.html). However, this example project specifically refers to the command-line interface tool [Arm Mbed CLI](https://github.com/ARMmbed/mbed-cli#installing-mbed-cli).
 
-## Import the example application
+## Application functionality
 
-From the command-line, import the example:
+The `main()` function outputs on the serial interface about a maximum of 8 threads information like `id`, `name`, `state`, `priority`, `stack_space`, `stack_size`.
```
Comment (evedon) [inline README.md]: Suggestion: 
The `main()` function outputs on the serial interface thread information such as `id`, `name`, `state`, `priority`, `stack_space`, `stack_size`.

**Review by evedon — CHANGES_REQUESTED** [review]
Looks good. A comment about starting user threads in main to show more information.

File: README.md [hunk 5]
```diff
@@ -72,7 +86,7 @@ The information below shows how to interpret the above fields:
 ```
 ID:            Thread id
 Name:          Thread name
-State:         Thread state
+State:         Thread states: `1` - Ready, `2` - Running, `3` - Waiting for delay to occur
```
Comment (evedon) [inline README.md]: The other states are missing:   
 osThreadInactive        =  0,         ///< Inactive.
  osThreadReady           =  1,         ///< Ready.
  osThreadRunning         =  2,         ///< Running.
  osThreadBlocked         =  3,         ///< Blocked.
  osThreadTerminated      =  4,         ///< Terminated.
  osThreadError           = -1,         ///< Error.

File: main.cpp [hunk 6]
```diff
@@ -21,12 +21,44 @@
...
+}
+
 int main()
 {
+    Thread *blinky_thread,
+           *idle_thread;    
     mbed_stats_thread_t *stats = new mbed_stats_thread_t[MAX_THREAD_STATS];
-    int count = mbed_stats_thread_get_each(stats, MAX_THREAD_STATS);
     
+    blinky_thread = new Thread(osPriorityNormal, MAX_THREAD_STACK, nullptr, "blinky_thread");
```
Comment (evedon) [inline main.cpp]: You should combine the declaration and definition:
```
Thread * blinky_thread = new Thread(osPriorityNormal, MAX_THREAD_STACK, nullptr, "blinky_thread");
```

File: main.cpp [hunk 7]
```diff
@@ -21,12 +21,44 @@
...
+    while(i--) { 
+        thread_sleep_for(wait_time_ms);
+    }
+}
+
 int main()
 {
+    Thread *blinky_thread,
+           *idle_thread;    
     mbed_stats_thread_t *stats = new mbed_stats_thread_t[MAX_THREAD_STATS];
```
Comment (evedon) [inline main.cpp]: Could you move this definition closer to where it is used, i.e. just before the call to `mbed_stats_thread_get_each`

File: main.cpp [hunk 8]
```diff
@@ -21,12 +21,44 @@
...
+    while(i--) {
+        led1 = !led1;
+        thread_sleep_for(wait_time_ms);
+    }
+}
+
+void idle()
+{
+    volatile uint64_t i = ~0;    
+    while(i--) {
```
Comment (evedon) [inline main.cpp]: Run astyle - this is not following coding style.

File: main.cpp [hunk 9]
```diff
@@ -21,12 +21,44 @@
...
  
 #define MAX_THREAD_STATS    0x8
- 
+
+// Initialise the digital pin LED1 as an output
+DigitalOut led1(LED1);
+
+#define MAX_THREAD_STACK 224
+
+int32_t wait_time_ms = 5000;
```
Comment (evedon) [inline main.cpp]: You could make this a `#define` instead. And blink every 0.5s or every s.

File: main.cpp [hunk 10]
```diff
@@ -21,12 +21,44 @@
...
+
+void blinky()
+{
+    volatile uint64_t i = ~0;
+    while(i--) {
```
Comment (evedon) [inline main.cpp]: Why not use `while(1)` to make the code simpler.

File: main.cpp [hunk 11]
```diff
@@ -21,12 +21,44 @@
...
+        led1 = !led1;
+        thread_sleep_for(wait_time_ms);
+    }
+}
+
+void idle()
+{
+    volatile uint64_t i = ~0;    
+    while(i--) { 
+        thread_sleep_for(wait_time_ms);
```
Comment (evedon) [inline main.cpp]: Perhaps you could have waited on a flag to show that this thread is blocked.

File: main.cpp [hunk 12]
```diff
@@ -21,12 +21,44 @@
 #endif
  
 #define MAX_THREAD_STATS    0x8
- 
+
+// Initialise the digital pin LED1 as an output
+DigitalOut led1(LED1);
+
+#define MAX_THREAD_STACK 224
```
Comment (evedon) [inline main.cpp]: Suggestion: rename to APP_THREAD_STACK

File: README.md [hunk 13]
```diff
@@ -1,59 +1,114 @@
...
-| mbed-os\rtos     | 13822 |  2310 | 4592 |
-| mbed-os\targets  |  9193 |   104 |  324 |
-| Subtotals        | 40397 |  2564 | 7260 |
-+------------------+-------+-------+------+
-Total Static RAM memory (data + bss): 9824 bytes
-Total Flash memory (text + data): 42961 bytes
+ID:            Thread id
+Name:          Thread name
+State:         Thread states:
+               osThreadInactive   =  0,
```
Comment (evedon) [inline README.md]: `State` is an unsigned integer, so you should just say `0: Thread is inactive` etc.

File: main.cpp [hunk 14]
```diff
@@ -21,13 +21,42 @@
 #endif
  
 #define MAX_THREAD_STATS    0x8
- 
+
+// Initialise the digital pin LED1 as an output
+DigitalOut led1(LED1);
+static EventFlags idle_ef;
+#define BLINKY_THREAD_STACK 224
```
Comment (evedon) [inline main.cpp:24]: Could you move all the `#define` in the same location, i.e. after `MAX_THREADS_STATS`?

File: README.md (hunk 13 above)
Comment (rajkan01) [reply README.md]: The headline is `Thread states`,  so we keep state alone as looks below:
```
State:         Thread states:
               0: Inactive,
               1: Ready,
               2: Running,
               3: Blocked,
               4: Terminated,
              -1: Error  
```

**Review by evedon — APPROVED** [review]

**Comment by evedon** [comment]
@AnotherButler This is ready for your review

**Comment by AnotherButler** [comment]
Could you please give me access to edit this repo?

**Review by AnotherButler — APPROVED** [review]
This looks fantastic 👍

**Pull request merged**